core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -340,6 +340,22 @@ class DAGScheduler(
     }
   }
 
+  /**
+   * Check to make sure we don't launch a barrier stage with unsupported RDD chain pattern. The
+   * following patterns are not supported:
+   * 1. Ancestor RDDs that have different number of partitions from the resulting RDD (eg.
+   * union()/coalesce()/first()/PartitionPruningRDD);
```

Comment:
but coalesce should be OK, right?  Is it just too fragile to allow coalesce while excluding the others?

Reply:
`coalesce()` is not safe when shuffle is false because it may cause the number of tasks doesn't match the number of partitions for the RDD that uses barrier mode.

Reply:
OK I see that it'll be a different number of partitions, but conceptually it should be OK, right?  the user just wants all tasks launched together, even if its a different number of tasks than the number of partitions in the original barrier rdd.

Reply:
but anyway, I guess its also fine to not support this case, I was just trying to understand myself.